## CarrerFoundry Task 4.8

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os


## Task 1

In [2]:
path = r'/Users/andrenedvidek/Desktop/Data Analaysis/CareerFoundry /Data Emersion /Task 4.3/02 Data copy/Prepared Data'

In [3]:
ords_prods_merge=pd.read_pickle(os.path.join(path,'ords_prods_merge_prepeard.pkl'))


## Task 2

In [4]:
df = ords_prods_merge

In [5]:
df.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  orders_day_of_week  ...  Gender  \
0  3139998.0    138.0    prior          28.0                 6.0  ...    Male   
1  1977647.0    138.0    prior          30.0                 6.0  ...    Male   
2   389851.0    709.0    prior           2.0                 0.0  ...  Female   
3   652770.0    764.0    prior           1.0                 3.0  ...  Female   
4  1813452.0    764.0    prior           3.0                 4.0  ...  Female   

       state   Age  date_joined nr_dependants family status    income  \
0  Minnesota  81.0     8/1/2019           1.0       married   49620.0   
1  Minnesota  81.0     8/1/2019           1.0       married   49620.0   
2    Vermont  66.0    6/16/2018           2.0       married  158302.0   
3  Wisconsin  40.0     2/9/2020           3.0       married   31308.0   
4  Wisconsin  40.0     2/9/2020           3.0       married   31308.0   

     price_range_loc     busiest_days      order_hour_of_day_loc  
0  Mid-range product   Regularly busy      busiest period of day  
1  Mid-range product   Regularly busy  medium busy period of day  
2  Mid-range product     Busiest days  medium busy period of day  
3  Mid-range product  Least busy days      busiest period of day  
4  Mid-range product  Least busy days  medium busy period of day  

[5 rows x 27 columns]

In [6]:
# calculate the mean of the “order_number” column grouped by the “department_id” column for the whole dataframe
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Task 3

### Actually the two results do differ quite a lot. Enough to reorganize which departments had the highest order number. For example, department 21 here has 22.9 as the average, while in the other subset, it has 25.5 as the average

## Task 4

In [7]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform('max')

In [8]:
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

     order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0   3139998.0    138.0    prior          28.0                 6.0  ...   
1   1977647.0    138.0    prior          30.0                 6.0  ...   
2    389851.0    709.0    prior           2.0                 0.0  ...   
3    652770.0    764.0    prior           1.0                 3.0  ...   
4   1813452.0    764.0    prior           3.0                 4.0  ...   
5   1701441.0    777.0    prior          16.0                 1.0  ...   
6   1871483.0    825.0    prior           3.0                 2.0  ...   
7   1290456.0    910.0    prior          12.0                 3.0  ...   
8    369558.0   1052.0    prior          10.0                 1.0  ...   
9    589712.0   1052.0    prior          15.0                 1.0  ...   
10  1798115.0   1379.0    prior           1.0                 5.0  ...   
11   497193.0   1480.0    prior          36.0                 3.0  ...   
12  1727826.0   1480.0    prior          38.0                 4.0  ...   
13  2694698.0   1480.0    prior          39.0                 2.0  ...   
14  3171753.0   1494.0    prior           6.0                 3.0  ...   

                   state   Age  date_joined  nr_dependants  \
0              Minnesota  81.0     8/1/2019            1.0   
1              Minnesota  81.0     8/1/2019            1.0   
2                Vermont  66.0    6/16/2018            2.0   
3              Wisconsin  40.0     2/9/2020            3.0   
4              Wisconsin  40.0     2/9/2020            3.0   
5                 Hawaii  51.0     1/4/2019            2.0   
6   District of Columbia  20.0     5/8/2019            3.0   
7              Tennessee  65.0    7/25/2019            0.0   
8                 Oregon  72.0    3/30/2018            1.0   
9                 Oregon  72.0    3/30/2018            1.0   
10                Alaska  61.0    3/30/2020            1.0   
11               Alabama  19.0   12/28/2019            1.0   
12               Alabama  19.0   12/28/2019            1.0   
13               Alabama  19.0   12/28/2019            1.0   
14               Indiana  78.0    6/29/2017            1.0   

                       family status    income    price_range_loc  \
0                            married   49620.0  Mid-range product   
1                            married   49620.0  Mid-range product   
2                            married  158302.0  Mid-range product   
3                            married   31308.0  Mid-range product   
4                            married   31308.0  Mid-range product   
5                            married   57797.0  Mid-range product   
6   living with parents and siblings   34171.0  Mid-range product   
7        

In [9]:
# Creating a flagged column 
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [10]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [11]:

ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [12]:
ords_prods_merge['loyalty_flag'].value_counts()

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
nan                       11
Name: count, dtype: int64

In [13]:
# looking at specific columns to check the data:
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0     138.0  Regular customer          28.0
1     138.0  Regular customer          30.0
2     709.0      New customer           2.0
3     764.0      New customer           1.0
4     764.0      New customer           3.0
5     777.0  Regular customer          16.0
6     825.0      New customer           3.0
7     910.0  Regular customer          12.0
8    1052.0  Regular customer          10.0
9    1052.0  Regular customer          15.0
10   1379.0      New customer           1.0
11   1480.0    Loyal customer          36.0
12   1480.0    Loyal customer          38.0
13   1480.0    Loyal customer          39.0
14   1494.0  Regular customer           6.0
15   1494.0  Regular customer           7.0
16   1494.0  Regular customer           9.0
17   1540.0  Regular customer           4.0
18   1540.0  Regular customer           5.0
19   1540.0  Regular customer           6.0
20   1540.0  Regular customer           7.0
21   1540.0  Regular customer           8.0
22   1540.0  Regular customer           9.0
23   1540.0  Regular customer          14.0
24   1540.0  Regular customer          15.0
25   1540.0  Regular customer          18.0
26   1540.0  Regular customer          19.0
27   1540.0  Regular customer          20.0
28   1540.0  Regular customer          21.0
29   1540.0  Regular customer          22.0
30   1540.0  Regular customer          24.0
31   1540.0  Regular customer          25.0
32   1540.0  Regular customer          27.0
33   1540.0  Regular customer          28.0
34   1598.0      New customer           1.0
35   1598.0      New customer           2.0
36   1598.0      New customer           3.0
37   1647.0      New customer           1.0
38   1647.0      New customer           4.0
39   1647.0      New customer           6.0
40   1647.0      New customer           8.0
41   1647.0      New customer           9.0
42   2833.0      New customer           1.0
43   2850.0  Regular customer           7.0
44   2857.0  Regular customer          24.0
45   3010.0  Regular customer           5.0
46   3029.0      New customer           1.0
47   3395.0      New customer           4.0
48   3904.0      New customer           1.0
49   4006.0    Loyal customer           2.0
50   4006.0    Loyal customer           3.0
51   4006.0    Loyal customer           5.0
52   4006.0    Loyal customer           9.0
53   4006.0    Loyal customer          12.0
54   4122.0  Regular customer          11.0
55   4141.0      New customer           2.0
56   4343.0      New customer           1.0
57   4553.0      New customer           3.0
58   4845.0  Regular customer           6.0
59   6866.0      New customer           1.0

## Task 5

In [14]:
ords_prods_merge.groupby('loyalty_flag').agg ({'prices':['mean','min', 'max', 'sum']})

prices                           
                       mean  min      max          sum
loyalty_flag                                          
Loyal customer    10.386336  1.0  99999.0  106814042.2
New customer      13.294670  1.0  99999.0   83011787.2
Regular customer  12.495717  1.0  99999.0  198391693.2
nan                8.327273  1.7     15.0         91.6

### The results are quite interesting. It shows that the new customers have the highest purchasing price average followed by regular customers and loyal customers are last. Overall, with the biggest purchasing group being regular customers, it can safely be said that this is where the most revenue is generated.

## Task 6

In [15]:
# Now, I need to create a new column named ave. This will help to group the customers into different spender classes.

ords_prods_merge['spend_ave'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [16]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0  3139998.0    138.0    prior          28.0                 6.0  ...   
1  1977647.0    138.0    prior          30.0                 6.0  ...   
2   389851.0    709.0    prior           2.0                 0.0  ...   
3   652770.0    764.0    prior           1.0                 3.0  ...   
4  1813452.0    764.0    prior           3.0                 4.0  ...   

   date_joined  nr_dependants  family status    income    price_range_loc  \
0     8/1/2019            1.0        married   49620.0  Mid-range product   
1     8/1/2019            1.0        married   49620.0  Mid-range product   
2    6/16/2018            2.0        married  158302.0  Mid-range product   
3     2/9/2020            3.0        married   31308.0  Mid-range product   
4     2/9/2020            3.0        married   31308.0  Mid-range product   

      busiest_days      order_hour_of_day_loc max_order      loyalty_flag  \
0   Regularly busy      busiest period of day      32.0  Regular customer   
1   Regularly busy  medium busy period of day      32.0  Regular customer   
2     Busiest days  medium busy period of day       5.0      New customer   
3  Least busy days      busiest period of day       3.0      New customer   
4  Least busy days  medium busy period of day       3.0      New customer   

   spend_ave  
0   6.935811  
1   6.935811  
2   7.930208  
3   4.972414  
4   4.972414  

[5 rows x 30 columns]

In [17]:
# Creating a flagged column for spender_typ 
ords_prods_merge.loc[ords_prods_merge['spend_ave'] < 10, 'spender_typ'] = str('low_spender')

In [18]:
ords_prods_merge.loc[ords_prods_merge['spend_ave'] >= 10, 'spender_typ'] = str('high_spender')

In [19]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0  3139998.0    138.0    prior          28.0                 6.0  ...   
1  1977647.0    138.0    prior          30.0                 6.0  ...   
2   389851.0    709.0    prior           2.0                 0.0  ...   
3   652770.0    764.0    prior           1.0                 3.0  ...   
4  1813452.0    764.0    prior           3.0                 4.0  ...   

   nr_dependants  family status    income    price_range_loc     busiest_days  \
0            1.0        married   49620.0  Mid-range product   Regularly busy   
1            1.0        married   49620.0  Mid-range product   Regularly busy   
2            2.0        married  158302.0  Mid-range product     Busiest days   
3            3.0        married   31308.0  Mid-range product  Least busy days   
4            3.0        married   31308.0  Mid-range product  Least busy days   

       order_hour_of_day_loc max_order      loyalty_flag spend_ave  \
0      busiest period of day      32.0  Regular customer  6.935811   
1  medium busy period of day      32.0  Regular customer  6.935811   
2  medium busy period of day       5.0      New customer  7.930208   
3      busiest period of day       3.0      New customer  4.972414   
4  medium busy period of day       3.0      New customer  4.972414   

   spender_typ  
0  low_spender  
1  low_spender  
2  low_spender  
3  low_spender  
4  low_spender  

[5 rows x 31 columns]

In [20]:
ords_prods_merge['spender_typ'].value_counts()

spender_typ
low_spender     31770614
high_spender      634245
nan                   11
Name: count, dtype: int64

## Task 7 

In [21]:
# Creating a column with the avergae of the days_since_prior_order by customer
ords_prods_merge['frequency_ave'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [22]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0  3139998.0    138.0    prior          28.0                 6.0  ...   
1  1977647.0    138.0    prior          30.0                 6.0  ...   
2   389851.0    709.0    prior           2.0                 0.0  ...   
3   652770.0    764.0    prior           1.0                 3.0  ...   
4  1813452.0    764.0    prior           3.0                 4.0  ...   

   family status    income    price_range_loc     busiest_days  \
0        married   49620.0  Mid-range product   Regularly busy   
1        married   49620.0  Mid-range product   Regularly busy   
2        married  158302.0  Mid-range product     Busiest days   
3        married   31308.0  Mid-range product  Least busy days   
4        married   31308.0  Mid-range product  Least busy days   

       order_hour_of_day_loc max_order      loyalty_flag spend_ave  \
0      busiest period of day      32.0  Regular customer  6.935811   
1  medium busy period of day      32.0  Regular customer  6.935811   
2  medium busy period of day       5.0      New customer  7.930208   
3      busiest period of day       3.0      New customer  4.972414   
4  medium busy period of day       3.0      New customer  4.972414   

   spender_typ  frequency_ave  
0  low_spender            8.0  
1  low_spender            8.0  
2  low_spender            8.0  
3  low_spender            9.0  
4  low_spender            9.0  

[5 rows x 32 columns]

In [23]:
#creating an additional Flag Column named frequency_user
ords_prods_merge.loc[ords_prods_merge['frequency_ave'] >20, 'customer_freq'] = str('Non-frequent customer')

In [24]:
ords_prods_merge.loc[(ords_prods_merge['frequency_ave'] > 10) & (ords_prods_merge['frequency_ave'] <= 20), 'customer_freq'] = str('Regular customer')

In [25]:
ords_prods_merge.loc[ords_prods_merge['frequency_ave'] <=10, 'customer_freq'] = str('Frequent customer')

In [26]:
ords_prods_merge[['user_id','frequency_ave', 'customer_freq', 'spend_ave','spender_typ']].head(60)

user_id  frequency_ave          customer_freq  spend_ave  spender_typ
0     138.0            8.0      Frequent customer   6.935811  low_spender
1     138.0            8.0      Frequent customer   6.935811  low_spender
2     709.0            8.0      Frequent customer   7.930208  low_spender
3     764.0            9.0      Frequent customer   4.972414  low_spender
4     764.0            9.0      Frequent customer   4.972414  low_spender
5     777.0           11.0       Regular customer   6.935398  low_spender
6     825.0           20.0       Regular customer   5.957576  low_spender
7     910.0            6.0      Frequent customer   6.680000  low_spender
8    1052.0           10.0      Frequent customer   7.162500  low_spender
9    1052.0           10.0      Frequent customer   7.162500  low_spender
10   1379.0           20.0       Regular customer   5.813636  low_spender
11   1480.0            7.0      Frequent customer   5.081633  low_spender
12   1480.0            7.0      Frequent customer   5.081633  low_spender
13   1480.0            7.0      Frequent customer   5.081633  low_spender
14   1494.0            7.0      Frequent customer   6.803797  low_spender
15   1494.0            7.0      Frequent customer   6.803797  low_spender
16   1494.0            7.0      Frequent customer   6.803797  low_spender
17   1540.0            7.0      Frequent customer   6.607122  low_spender
18   1540.0            7.0      Frequent customer   6.607122  low_spender
19   1540.0            7.0      Frequent customer   6.607122  low_spender
20   1540.0            7.0      Frequent customer   6.607122  low_spender
21   1540.0            7.0      Frequent customer   6.607122  low_spender
22   1540.0            7.0      Frequent customer   6.607122  low_spender
23   1540.0            7.0      Frequent customer   6.607122  low_spender
24   1540.0            7.0      Frequent customer   6.607122  low_spender
25   1540.0            7.0      Frequent customer   6.607122  low_spender
26   1540.0            7.0      Frequent customer   6.607122  low_spender
27   1540.0            7.0      Frequent customer   6.607122  low_spender
28   1540.0            7.0      Frequent customer   6.607122  low_spender
29   1540.0            7.0      Frequent customer   6.607122  low_spender
30   1540.0            7.0      Frequent customer   6.607122  low_spender
31   1540.0            7.0      Frequent customer   6.607122  low_spender
32   1540.0            7.0      Frequent customer   6.607122  low_spender
33   1540.0            7.0      Frequent customer   6.607122  low_spender
34   1598.0            2.0      Frequent customer   7.123529  low_spender
35   1598.0            2.0      Frequent customer   7.123529  low_spender
36   1598.0            2.0      Frequent customer   7.123529  low_spender
37   1647.0           25.0  Non-frequent customer   4.853521  low_spender
38   1647.0           25.0  Non-frequent customer   4.853521  low_spender
39   1647.0           25.0  Non-frequent customer   4.853521  low_spender
40   1647.0           25.0  Non-frequent customer   4.853521  low_spender
41   1647.0           25.0  Non-frequent customer   4.853521  low_spender
42   2833.0           13.0       Regular customer   7.377143  low_spender
43   2850.0            7.0      Frequent customer   8.054153  low_spender
44   2857.0           11.0       Regular customer   8.353793  low_spender
45   3010.0            7.0      Frequent customer   6.452800  low_spender
46   3029.0           28.0  Non-frequent customer   7.960606  low_spender
47   3395.0           18.0       Regular customer   6.322222  low_spender
48   3904.0           21.0  Non-frequent customer   7.511111  low_spender
49   4006.0            7.0      Frequent customer   7.441100  low_spender
50   4006.0            7.0      Frequent customer   7.441100  low_spender
51   4006.0            7.0      Frequent customer   7.441100  low_spender
52   4006.0            7.0      Frequent customer   7.441100  low_spender
53   400

## Task 9 

In [27]:
ords_prods_merge.to_pickle(os.path.join(path,'ords_prods_merge_prepeard_2.pkl'))